### ISSUE TITLE
Pyston for macOS

### ISSUE THREAD

**Issue by mafrosis** [pyston/pyston_v2#16]
https://github.com/pyston/pyston_v1/issues/165

Interested to know if anyone has thought about this recently.

**Comment by kmod** [comment]
Should be much easier than Pyston v1, but since I don't think either of us have Mac build experience we could use some help

**Comment by bferguson3** [comment]
I can try to help but I am unfamiliar with your build chain, as I just discovered pyston. 
First errors appear to be lexical within the makefile:

make: lsb_release: Command not found
make: *** No rule to make target `pyston/build/cpython_opt_install/usr/bin/python3.bolt', needed by `pyston/build/opt_env/bin/python'.  Stop.

**Comment by xqzhou** [comment]
I tried to build on mac but failed.

Anyone succeeded building on macOS?

**Comment by mischievous** [comment]
How do I generate the aot.h since its in the git .ignore files?


First error i'm getting from a the released code... 
make: lsb_release: Command not found
cat: /proc/cpuinfo: No such file or directory
make: lsb_release: Command not found
make: *** No rule to make target `build/opt_env/bin/python', needed by `pyston3'.  Stop.


Next: "./configure --enable-configure" to rebuild the Makefile for osx.  I'm getting this error "make: *** No rule to make target `../aot/aot.h', needed by `Python/aot_ceval.o'.  Stop." with this file ./Python/dynload_shlib.c.   aot.h is an ignored file.  Not seeing how it gets built.  

If I disable the AOT -- "./configure --enable-configure --disable-aot" it then fails with 
Undefined symbols for architecture x86_64:
  "_PyInit_aot_ceval", referenced from:
      _pyinit_main in libpython3.8-pyston2.3a.a(pylifecycle.o)
  "_aot_exit", referenced from:
      _Py_FinalizeEx in libpython3.8-pyston2.3a.a(pylifecycle.o)
ld: symbol(s) not found for architecture x86_64

**Comment by kmod** [comment]
The reason we added the `--enable-configure` flag is because it no longer works to call configure directly :) We do not use a standard `./configure && make` build process because we need to build CPython multiple times with different configure args, so we have a top-level Makefile which drives the whole process (this Makefile is not generated by configure).

To answer your specific question, aot.h is generated in parallel with aot_profile.c (another file that the CPython Makefile references but doesn't build itself) which is why it doesn't appear in our Makefile.

I'm surprised by those undefined symbols, is it possible that you have build artifacts from the non-disable-aot build?

But anyway the real answer is to start from our Makefile; a simple `make` (or `make unopt` for faster iteration) will set you down the path of all the build rules we've written.

Let us know if we can help at all!

**Comment by mischievous** [comment]
Hello, thanks for getting back to me.  

Used the released 2.3.1 tar.gz and a clone of the master repo these are the errors i'm getting... 

Demon:pyston-pyston_2.3.1 alexandra$ make -j 2
make: lsb_release: Command not found
cat: /proc/cpuinfo: No such file or directory
make: lsb_release: Command not found
make: *** No rule to make target `build/opt_env/bin/python', needed by `pyston3'.  Stop.

1)  MacOS does not have a "proc" directory but it looks like make is ignoring this error
2) I'm not sure where the build directory is getting created, or populated with python3.8

**Comment by kmod** [comment]
What happens if you do `make unopt -j2`? That target is simpler so it would be easier to start with.

`build/opt_env/bin/python` is generated by [this line](https://github.com/pyston/pyston/blob/pyston_main/Makefile#L145) which is called from [this line](https://github.com/pyston/pyston/blob/pyston_main/Makefile#L216). You can try inserting something like `$(info $(1))` before line 145 to see if it's being called correctly.

There's going to be a lot of Makefile spelunking to do, it might be worth taking a read through the Makefile to familiarize yourself first :)

**Comment by mischievous** [comment]
I would be happy to go spelunking but I've avoided any complex makefiles for my entire life.  Makefiles and me don't get along.  


With the unopt... and the "$(info $(1))" at line 145

Demon:pyston alexandra$ make unopt -j2
make: lsb_release: Command not found
cat: /proc/cpuinfo: No such file or directory
unopt
unoptshared
opt
optshared
releaseunopt
releaseunoptshared
release
releaseshared
dbg
dbgshared
stock
stockunopt
stockdbg
make: lsb_release: Command not found
make: *** No rule to make target `unopt'.  Stop.

**Comment by kmod** [comment]
Perhaps MacOS `make` is different from gnu make? There is some basic functionality of our Makefile that is not working for you, and I don't know why. I'm happy to answer any questions, but I don't think I can do the debugging remotely.

**Comment by mischievous** [comment]
Ratz!  Ok.  Should I try with gnu_make?

**Comment by kmod** [comment]
Oh there's a separate `gnu_make`? That definitely sounds like a good thing to try, we haven't tested our makefile with any other make.

**Comment by mischievous** [comment]
I think thats the issue... I also tyied to build it on a AWS linux workspace.  It gave me the same error.  Then I thought I could install gnu make but turns out I can install gnustep-make.  

So I will have to see about getting gnu make made.  

I'll keep you informed.

**Comment by kmod** [comment]
If you're getting an error on a linux machine then maybe there's something else that is going wrong. Maybe try to get the build working at all first (we test the most on Ubuntu) to make sure your setup is good?

I noticed that you said you used the .tar.gz but that shouldn't be necessary for the build, how did you end up using it?

**Comment by undingen** [comment]
Also one issue you may run into later on: the JIT helper functions (`Python/aot_ceval_jit_helper.c`) use a extension which only gcc suports so clang will not be enough to compile pyston :/.

**Comment by mischievous** [comment]
On the AWS workspace the gnu make is doing more.  Having to update to the newest version of cmake and not the version with AWS Linux. 

It appears to be building.  Since I used -j2 it might take some time...

**Comment by mischievous** [comment]
Well back to the Mac... I've installed gnu make, and made sure cmake was up to date.  I Went into the makefile and removed the lines dealing with lsb_release, and /proc/cpuinfo and had them perform the actions that would have taken that matches the AWS linux i'm also testing with.  

It still acts like gnu make is not doing the $(1) replacement correctly in the targets.


I'm going to try one more thing if I can get the AWS built correctly.  


Demon:pyston alexandra$ make -v
GNU Make 3.81
Copyright (C) 2006  Free Software Foundation, Inc.
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.

This program built for i386-apple-darwin11.3.0

Demon:pyston alexandra$ make -j2
make: *** No rule to make target `build/opt_env/bin/python', needed by `pyston3'.  Stop.

**Comment by mischievous** [comment]
The AWS linux built... 

1. download/build gnu make
2. download/build cmake
3. pip install virtualenv (does not install via the yum install like apt-get does I guess). 
4. make -j2 worked... 
    a. make unopt -j2 did not work. 

For the Mac I was hoping I could just run make with --dry-run to see everything that is being built.  But that will not work.  :(

**Comment by kmod** [comment]
Ah very weird, keep us updated if you find any clues on why this is happening!

**Comment by mischievous** [comment]
What version of llvm/clang are you using?

**Comment by undingen** [comment]
The submodule of llvm and clang we include and build is v10. The Makefile uses the system supplied clang to build it and on my machine (ubuntu 20.04) it's also v10 but it should work with a different version too.

**Comment by mischievous** [comment]
You would think it would.  The the system default is the version from Apple which is failing (can't remember the failure right now) but I downloaded version 11 and put it into my path.  It's complaining about "-fure-ld" option not found.  

So I will get rid of 11 and try 10 now... 

Thanks... Will keep you informed. 


/usr/local/clang+llvm-11.0.0-x86_64-apple-darwin/bin/clang++  -DCXX_SUPPORTS_CUSTOM_LINKER  -Werror=unguarded-availability-new -fuse-ld=gold -isysroot /Volumes/Development/Xcode-11.3.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk -Wl,-search_paths_first -Wl,-headerpad_max_install_names CMakeFiles/cmTC_07721.dir/src.cxx.o -o cmTC_07721  -lm
clang-11: error: invalid linker name in argument '-fuse-ld=gold'

**Comment by undingen** [comment]
mmh I guess it can't find the gold linker.  I think it's ELF only so not available on Mac. Please try removing `-DLLVM_USE_LINKER=gold`

**Comment by mischievous** [comment]
Trying now... Like you suspected 10 was no better.  :D

**Comment by mischievous** [comment]
So the LLVM/CLANG from https://releases.llvm.org/download.html is not working so well.  Removing the "-DLLVM_USE_LINKER=gold" removes the errors until it gets to the "checking whether the C compiler works..." where it just hangs. 


mkdir -p build/bc_build
cd build/bc_build; WRAPPER_REALCC=/Volumes/Promise/test/pyston/pyston/build/Release/llvm/bin/clang-10 WRAPPER_OUTPUT_PREFIX=../cpython_bc CC=../../pyston/clang_wrapper.py CFLAGS_NODIST="-fstack-protector -specs=/Volumes/Promise/test/pyston/pyston/pyston/tools/no-pie-compile.specs -D_FORTIFY_SOURCE=2 -Wno-unused-command-line-argument" LDFLAGS_NODIST="-specs=/Volumes/Promise/test/pyston/pyston/pyston/tools/no-pie-link.specs -Wl,-z,relro" ../../configure --prefix=/usr --disable-aot --disable-debugging-features --enable-configure
checking for git... found
checking build system type... x86_64-apple-darwin19.6.0
checking host system type... x86_64-apple-darwin19.6.0
checking for pyston2.3... no
checking for pyston3... no
checking for pyston... no
checking for python2.3... no
checking for python3... python3
checking for --enable-universalsdk... no
checking for --with-universal-archs... no
checking MACHDEP... "darwin"
checking for gcc... ../../pyston/clang_wrapper.py
checking whether the C compiler works...

**Comment by tzekid-nd** [comment]
@mischievous did you manage to get it running in the end / make any further progress?

**Comment by undingen** [comment]
We changed now our Makefile to optionally use installed clang+llvm(+bolt) binaries when `PYSTON_USE_SYS_BINS=1` is defined.
This may make it also quicker to start developing so maybe worth a try. It's important that the clang and llvm used is v10 so the paths in the first we lines of the Makefile may need changes.

**Comment by mischievous** [comment]
Nope.   Sorry... My configure/make foo is too weak.  I can give it a try with PYSTON_USE_SYS_BINS=1 and see if it works any better.

**Comment by tzekid-nd** [comment]
Running make on macos aarch64 ( m1 ) gave the following err:  
Ran on the main branch... there seems to be no rule for 'build/opt_env/bin/python' or 'unopt' in the makefile.  
Did I miss something?  

```bash
$ PYSTON_USE_SYS_BINS=1 make -j 4
make: *** No rule to make target `build/opt_env/bin/python', needed by `pyston3'.  Stop.
```

**Comment by kmod** [comment]
From an above comment it sounded like we may be relying on gnu-make-specific features, could you try using that make instead?

**Comment by mischievous** [comment]
Yes you have to use gnu make.  Also don't forget to remove the "-DLLVM_USE_LINKER=gold"

**Comment by Calinou** [comment]
If anyone is stumbling upon this issue from a web search, remember that Pyston Lite works on macOS (both x86 and ARM): https://blog.pyston.org/2022/06/08/announcing-pyston-lite-our-python-jit-as-an-extension-module/

> As a bonus, this is the first time that Mac users can use the Pyston JIT.

It won't give you all the performance benefits of full Pyston, but it's easier to install and manage.
